File: hello.go
/*
  tools for check server:port available
  author: smallfish
*/

package main

import (
    "net"
    "fmt"
    "time"
    "flag"
    "strings"
    "sync"
    "os"
    "io/ioutil"
)

var (
    file        string
    limit       int
    timeout     int
    wg          sync.WaitGroup
    ch          chan int
)

func init() {
    flag.StringVar(&file, "file", "", "host:port file")
    flag.IntVar(&limit, "limit", 10, "limit concurrency, default: 10")
    flag.IntVar(&timeout, "timeout", 5, "connect timeout, default: 5")
}

func getAddrs(file string) ([]string) {
    byte, err := ioutil.ReadFile(file)
    if err != nil {
        return nil
    }
    var ret []string
    for _, val := range strings.Split(string(byte), "\n") {
        if len(val) > 2 {
            ret = append(ret, val)
        }
    }
    return ret
}

func work(addr string) {
    ch <- 1
    conn, err := net.DialTimeout("tcp", addr, time.Duration(timeout) * time.Second)
    if err != nil {
        fmt.Printf("%s\t\033[31mERROR\033[0m\n", addr) // red ERROR
    } else {
        conn.Close()
        fmt.Printf("%s\t\033[32mOK\033[0m\n", addr) // green OK
    }
    wg.Done()
    <-ch
}

func main() {
    flag.Usage = func() { flag.PrintDefaults() }
    flag.Parse()
    if file == "" {
        flag.PrintDefaults()
        os.Exit(0)
    }
    ch = make(chan int, limit)
    for _, addr := range getAddrs(file) {
        wg.Add(1)
        go work(addr)
    }
    wg.Wait()
}

File: hello.py
"""
  tools for check server:port available, require gevent(use coroutine)
  author: smallfish
"""
from sys import exit
from optparse import OptionParser
from itertools import izip_longest
import gevent
from gevent import socket

def gen_groups(array, n):
    args = [iter(array)] * n
    return list(izip_longest(*args))

def get_connected(addr, port):
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.connect((addr, port))
        sock.close()
        return addr, port, True
    except:
        return addr, port, False

def main():
    parser = OptionParser(usage='%prog --file=urlfile --limit=1000 --timeout=5')
    parser.add_option('', '--file', help='host:port file')
    parser.add_option('', '--limit', default=10, type="int", help='limit concurrency, default: %default')
    parser.add_option('', '--timeout',  default=5, type="int", help='connect timeout, default: %default')
    (options, args) = parser.parse_args()
    if not options.file:
        parser.print_help()
        exit(0)
    socket.setdefaulttimeout(options.timeout)
    groups = gen_groups([line.strip() for line in open(options.file) if line], options.limit)
    for group in groups:
        jobs = []
        for item in group:
            if item:
                addr, port = item.split(":")
                jobs.append(gevent.spawn(get_connected, addr, int(port)))
        gevent.joinall(jobs)
        for job in jobs:
            addr, port, connected = job.value
            if connected:
                print "%s:%d\tOK" % (addr, port)
            else:
                print "%s:%d\tERROR" % (addr, port)

if __name__ == '__main__':
    main()
